File: bsev.pl
package Programme;
use Moo;
use Types::Standard qw/ Maybe Str /;

has id => ( is => 'ro', isa => Maybe[Str] );
has id2 => ( is => 'ro', isa => Maybe[Str] );

package Brand;
use Moo;
extends 'Programme';

package Series;
use Moo;
extends 'Programme';

package Episode;
use Moo;
extends 'Programme';

package Database;
use Moo;
use MooX::HandlesVia;
use Types::Standard qw/ ArrayRef HashRef InstanceOf /;

use experimental 'signatures', 'lexical_subs';

has list => (
    is => 'ro',
    isa => ArrayRef[InstanceOf["Programme"]],
    handles_via => 'Array',
    handles => {
        filter => 'grep',
    }
);

has graph => (
    is => 'ro',
    isa => HashRef[InstanceOf["Programme"]],
    handles_via => 'Hash',
    handles => {
        lookup => 'get'
    }
);

sub search ($self, $prog1) {
    my sub match ($prog2) {
        return unless ref $prog1 eq ref $prog2;

        return unless $prog1->id or $prog1->id2;
        return unless $prog2->id or $prog2->id2;
        
        return unless $prog1->id or $prog2->id2;
        return unless $prog1->id2 or $prog2->id;

        return if $prog1->id and $prog2->id 
            and $prog1->id ne $prog2->id;
        return if $prog1->id2 and $prog2->id2 
            and $prog1->id2 ne $prog2->id2;

        return 1;
    }

    return $self->filter(\&match);
}

sub getHierarchy($self, $prog) {
    if (my $parent = $self->lookup($prog->id)) {
        return ($prog, $self->getHierarchy($parent))
    }
    else {
        return $prog;
    }

}

package main;
use Data::Dumper::Concise;

my $b1 = Brand->new( id => "bbc" );
my $s1 = Series->new( id => "hignfy");
my $e1 = Episode->new( id => "hignfy1");

my $db = Database->new(
    list => [$b1, $s1, $e1],
    graph => {
        $e1->id => $s1,
        $s1->id => $b1,
    }
);

print Dumper(
    $db->getHierarchy(
        $db->search(
            Episode->new( id => 'hignfy1' )
        )
    )
);

File: bsev.hs

import qualified Data.Map as M

type PrimaryId = Maybe String
type AlternateId = Maybe String

data Prog = Prog ProgType PrimaryId AlternateId
    deriving (Eq, Ord, Show)

data ProgType = Brand | Series | Episode
    deriving (Eq, Ord, Show)

data Database = Database {
    list  :: [Prog],
    graph :: M.Map Prog Prog
}

search :: Database -> Prog -> [Prog]
search db p =
    filter (match p) (list db)
    where
        match (Prog t1 _ _) (Prog t2 _ _) | t1 /= t2 = False

        match (Prog _ Nothing Nothing) _ = False
        match _ (Prog _ Nothing Nothing) = False

        match (Prog _ Nothing _) (Prog _ _ Nothing) = False
        match (Prog _ _ Nothing) (Prog _ Nothing _) = False

        match (Prog _ (Just p1) _) (Prog _ (Just p2) _) | p1 /= p2 = False
        match (Prog _ _ (Just a1)) (Prog _ _ (Just a2)) | a1 /= a2 = False

        match _ _ = True

getHierarchy :: Database -> Prog -> [Prog]
getHierarchy db p =
    let l = M.lookup p (graph db)
    in case l of
        Nothing -> [p]
        Just (parent) -> p : (getHierarchy db parent)

-- sample data and call

b1 = Prog Brand (Just "bbc1") Nothing
s1 = Prog Series (Just "hignfy") Nothing
e1 = Prog Episode (Just "hignfy1") (Just "hignfy1a")

db :: Database
db = Database [b1, s1, e1] 
    (M.fromList [ 
        (e1, s1),
        (s1, b1)
    ])

main = do
    print $ search db (Prog Episode (Just "hignfy1") Nothing) >>= getHierarchy db
    print $ search db (Prog Episode Nothing (Just "hignfy1a")) >>= getHierarchy db
    print $ search db (Prog Series (Just "hignfy") (Just "new ID")) >>= getHierarchy db
